### PR TITLE
Fix #8001: Display the VPN subscription in the App Store

### LINKED ISSUE #8001: Display the VPN subscription in the App Store
### Details - Setting up Promotion and Implementation

To promote in app purchase on AppStore these steps should be handled on AppStore Connect side as 

`Step 1`

You can promote in-app purchases and subscriptions on your App Store product page if you’ve [added a promotional image](https://developer.apple.com/help/app-store-connect/create-an-app-record/view-and-edit-app-information)

After that these steps should be followed.

- In-app purchase should be Add or remove in-app purchases for promotion

- In-app purchase should be Customize the order of your promoted in-app purchases

- Promote only relevant in-app purchases to customers ) this step will also need code changes using [SKProductStorePromotionController](https://developer.apple.com/documentation/storekit/skproductstorepromotioncontroller)

So after AppStore Connect side is added with proper promotional image and other criteria, next step should be adding proper delegate method into the source code.

https://developer.apple.com/help/app-store-connect/configure-in-app-purchase-settings/promote-in-app-purchases

`Step 2`

- Adding delegate method shouldAddStorePayment 
[paymentQueue(_:shouldAddStorePayment:for:)](https://developer.apple.com/documentation/storekit/skpaymenttransactionobserver/2877502-paymentqueue) delegate method from SKPaymentTransactionObserver should be implemented 

This protocol method should determine if payment can be processed if user pressed the promotion on AppStore and opened the app.

This method's implementation should handle important steps of recognizing App is currently on a state(like onboarding) and return status accordingly for payment process, If not save the deferred payment until payment can be processed in a better situation.

In addition it will signal main browser to show VPN Buy Screen before triggering the payment to make user have the context before buying or providing a second chance to make them think about the deal.

- Properly Handle Visibility Settings 
[SKProductStorePromotionController](https://developer.apple.com/documentation/storekit/skproductstorepromotioncontroller) will be implemented to handle the proper visibility and order for the purchases, on top of the order visibility assigned in AppStore Connect to have specific mobile device criteria such as If the in-app Purchase already implemented not to show that to user. 

https://developer.apple.com/documentation/storekit/skpaymenttransactionobserver/promoting_in-app_purchases

### Testing

Testing Subscription Promotion is not very well designed on Apple side only way to test it is running promotion URL in another browser like Safari which in our case will be itms-services://?action=purchaseIntent&bundleId=com.brave.ios.browser&productIdentifier=bravevpn.monthly if this is a Release build for monthly transaction in app purchase

https://developer.apple.com/documentation/storekit/in-app_purchase/testing_promoted_in-app_purchases

### PR DESCRIPTION
**Pull request by soner-yuksel**
## Summary of Changes

This pull request fixes #8001

Please refer to the issue for details about implementaton, promotion AppStore Connect angle and Testing and details for all of them.

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New or updated UI has been tested across:
  - [x] Light & dark mode
  - [x] Different size classes (iPhone, landscape, iPad)
  - [x] Different dynamic type sizes

## Test Plan:

- Install the app and finish onboarding
- Kill and re-run the app
- Trigger promotional purchase link 
For example:[ itms-services://action=purchaseIntent&bundleId=com.brave.ios.browser&productIdentifier=bravevpn.monthly](itms-services://action=purchaseIntent&bundleId=com.brave.ios.browser&productIdentifier=bravevpn.monthly)
in Release scheme
- Check monthly payment is triggered

## Screenshots:

https://github.com/brave/brave-ios/assets/6643505/6e69be95-3e80-4800-9c0c-bb692d4fa516

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Sources/Brave/Frontend/Browser/BrowserViewController.swift [hunk 1]
```diff
@@ -3344,3 +3351,22 @@ extension BrowserViewController {
...
+    // No-op
+  }
+  
+  public func purchaseFailed(error: IAPObserver.PurchaseError) {
+    // No-op
+  }
+  
+  public func handlePromotedInAppPurchase() {
+    // Open VPN Buy Screen before system triggers buy action
+    // Delaying the VPN Screen launch delibrately to syncronize promoted purcahse launch
```
Comment (iccub) [inline Sources/Brave/Frontend/Browser/BrowserViewController.swift]: `purcahse` typo

File: Sources/BraveVPN/BraveVPN.swift [hunk 2]
```diff
@@ -295,6 +295,29 @@ public class BraveVPN {
     helper.mainCredential?.hostnameDisplayValue
   }
 
+  /// Type of vpnsubscription
```
Comment (iccub) [inline Sources/BraveVPN/BraveVPN.swift]: `vpnsubscription` space

File: Sources/BraveVPN/BraveVPN.swift [hunk 3]
```diff
@@ -650,6 +673,66 @@ public class BraveVPN {
     }
   }
   
+  // MARK: - Promotion
+  
+  /// Editing product promotion order first monthly and yearly after
```
Comment (iccub) [inline Sources/BraveVPN/BraveVPN.swift]: Didn't they ask to show the longer option first? I'd need to look at the docs again

**Review by iccub — CHANGES_REQUESTED** [review]
as per offline convo, please do not show the promo if the user bought using the Brave website

File: Sources/BraveVPN/BraveVPN.swift (hunk 3 above)
Comment (soner-yuksel) [reply Sources/BraveVPN/BraveVPN.swift]: Fixed under https://github.com/brave/brave-ios/pull/8415/commits/2ffc995bf5e7d4b042512f8eaa9e17fb58588cff

File: Sources/Brave/Frontend/Browser/BrowserViewController.swift (hunk 1 above)
Comment (soner-yuksel) [reply Sources/Brave/Frontend/Browser/BrowserViewController.swift]: Fixed at https://github.com/brave/brave-ios/pull/8415/commits/b59fc88c2ee0df156768a4574e2438d50289cfcf

File: Sources/BraveVPN/BraveVPN.swift (hunk 2 above)
Comment (soner-yuksel) [reply Sources/BraveVPN/BraveVPN.swift]: Fixed at https://github.com/brave/brave-ios/pull/8415/commits/b59fc88c2ee0df156768a4574e2438d50289cfcf

**Comment by soner-yuksel** [comment]
> as per offline convo, please do not show the promo if the user bought using the Brave website

The case for purchase se on website handled https://github.com/brave/brave-ios/pull/8415/commits/b59fc88c2ee0df156768a4574e2438d50289cfcf

**Review by iccub — APPROVED** [review]
very nice :)

**Comment by iccub** [comment]
tests need a fix

**Pull request merged**